flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/HITSAlgorithm.java

```diff
@@ -0,0 +1,178 @@
+import org.apache.flink.graph.EdgeDirection;
+import org.apache.flink.graph.Graph;
+import org.apache.flink.graph.GraphAlgorithm;
+import org.apache.flink.graph.Vertex;
+import org.apache.flink.graph.spargel.MessageIterator;
+import org.apache.flink.graph.spargel.MessagingFunction;
+import org.apache.flink.graph.spargel.ScatterGatherConfiguration;
+import org.apache.flink.graph.spargel.VertexUpdateFunction;
+import org.apache.flink.types.DoubleValue;
+import org.apache.flink.util.Preconditions;
+
+/**
+ * This is an implementation of HITS algorithm, using a scatter-gather iteration.
+ * The user can define the maximum number of iterations. HITS algorithm is determined by two parameters,
+ * hubs and authorities. A good hub represented a page that pointed to many other pages, and a good authority
+ * represented a page that was linked by many different hubs. The implementation assumes that the two value on
+ * every vertex are the same at the beginning.
+ * <p>
+ * If the number of vertices of the input graph is known, it should be provided as a parameter
+ * to speed up computation. Otherwise, the algorithm will first execute a job to count the vertices.
+ */
+public class HITSAlgorithm<K, EV> implements GraphAlgorithm<K, Double, EV, DataSet<Vertex<K, Double>>> {
+
+	public static enum HITSParameter {
+		HUB,
+		AUTHORITY
+	}
+
+	private int maxIterations;
+	private long numberOfVertices;
+
+	/**
+	 * Creates an instance of HITS algorithm.
+	 * If the number of vertices of the input graph is known,
+	 * use the {@link HITSAlgorithm#HITSAlgorithm(int, long, HITSParameter)} constructor instead.
+	 *
+	 * @param maxIterations the maximum number of iterations
+	 * @param hitsParameter the type of final web pages users want to get by this algorithm
+	 */
+	public HITSAlgorithm(int maxIterations, HITSParameter hitsParameter) {
+		Preconditions.checkArgument(maxIterations > 0, "The number of maximum iteration should be greater than 0.");
+		if (hitsParameter == HITSParameter.AUTHORITY) {
+			this.maxIterations = maxIterations * 2;
+		} else {
+			this.maxIterations = maxIterations * 2 + 1;
+		}
+	}
+
+	/**
+	 * Creates an instance of HITS algorithm.
+	 * If the number of vertices of the input graph is unknown,
+	 * use the {@link HITSAlgorithm#HITSAlgorithm(int, HITSParameter)} constructor instead.
+	 *
+	 * @param maxIterations the maximum number of iterations
+	 * @param hitsParameter the type of final web pages users want to get by this algorithm
+	 */
+	public HITSAlgorithm(int maxIterations, long numberOfVertices, HITSParameter hitsParameter) {
+		this(maxIterations, hitsParameter);
+		Preconditions.checkArgument(numberOfVertices > 0, "The number of vertices in graph should be greater than 0.");
+		this.numberOfVertices = numberOfVertices;
+	}
+
+	@Override
+	public DataSet<Vertex<K, Double>> run(Graph<K, Double, EV> netGraph) throws Exception {
+		if (this.numberOfVertices == 0) {
+			this.numberOfVertices = netGraph.numberOfVertices();
```

Comment:
I'm sorry that i forgot to remove it. `Sum normalization` does not need the number of vertices, only necessary for `z-core normalization`. And i have submitted a new commit for this.